File: EdgeSecurity-v6.3/app/backend/services/risk_engine.py
"""Motor inicial de risco para aproximação humano-máquina.

IMPORTANTE: a distância usada aqui é baseada em pixels/caixas. Isso é útil para
protótipo, mas NÃO representa distância física em metros. Para uso industrial,
o sistema deve ser calibrado para cada câmera (homografia, zonas físicas,
profundidade/estéreo ou outro método validado) antes de ser usado como barreira
de segurança.
"""
from math import hypot


def center(box):
    x1, y1, x2, y2 = box
    return ((x1 + x2) / 2, (y1 + y2) / 2)


def box_gap(a, b):
    ax1, ay1, ax2, ay2 = a
    bx1, by1, bx2, by2 = b
    dx = max(ax1 - bx2, bx1 - ax2, 0)
    dy = max(ay1 - by2, by1 - ay2, 0)
    return hypot(dx, dy)


def assess_risk(detections):
    people = [d for d in detections if d["class_name"] == "human"]
    machines = [d for d in detections if d["class_name"] in {"forklift", "machine"}]
    risks = []

    for person in people:
        for machine in machines:
            gap = box_gap(person["bbox"], machine["bbox"])
            pc = center(person["bbox"])
            mc = center(machine["bbox"])
            center_distance = hypot(pc[0] - mc[0], pc[1] - mc[1])

            # Faixas conservadoras para o protótipo visual. Devem ser calibradas.
            if gap <= 0:
                level = "critical"
            elif gap <= 40:
                level = "high"
            elif gap <= 90:
                level = "medium"
            else:
                level = "safe"

            risks.append({
                "person_track_id": person.get("track_id"),
                "machine_track_id": machine.get("track_id"),
                "machine_class": machine.get("class_name", "machine"),
                "machine_label": machine.get("label", "machine"),
                "gap_pixels": round(gap, 1),
                "center_distance_pixels": round(center_distance, 1),
                "level": level,
            })

    priority = {"critical": 4, "high": 3, "medium": 2, "safe": 1}
    overall = max((r["level"] for r in risks), key=lambda x: priority[x], default="safe")
    return {"level": overall, "pairs": risks}

File: js/risk-engine.js
window.RiskEngine = (() => {
  function center(b) {
    return [(b[0] + b[2]) / 2, (b[1] + b[3]) / 2];
  }
  function boxGap(a, b) {
    const dx = Math.max(a[0] - b[2], b[0] - a[2], 0);
    const dy = Math.max(a[1] - b[3], b[1] - a[3], 0);
    return Math.hypot(dx, dy);
  }
  function assess(detections) {
    const people = detections.filter((d) => d.class_name === "human");
    const machines = detections.filter((d) => d.class_name === "forklift" || d.class_name === "machine");
    const risks = [];
    for (const p of people) {
      for (const m of machines) {
        const gap = boxGap(p.bbox, m.bbox);
        const pc = center(p.bbox);
        const mc = center(m.bbox);
        const cd = Math.hypot(pc[0] - mc[0], pc[1] - mc[1]);
        let level = "safe";
        if (gap <= 0) level = "critical";
        else if (gap <= 40) level = "high";
        else if (gap <= 90) level = "medium";
        risks.push({ person_track_id: p.track_id ?? null, machine_track_id: m.track_id ?? null, gap_pixels: Math.round(gap * 10) / 10, center_distance_pixels: Math.round(cd * 10) / 10, level });
      }
    }
    const prio = { critical: 4, high: 3, medium: 2, safe: 1 };
    let overall = "safe";
    let best = 0;
    for (const r of risks) if ((prio[r.level] || 0) > best) { best = prio[r.level]; overall = r.level; }
    return { level: overall, pairs: risks };
  }
  return { assess, center, boxGap };
})();
